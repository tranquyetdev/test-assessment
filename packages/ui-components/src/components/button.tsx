import * as React from 'react';

export interface ButtonProps {
  onClick?: () => void;
  children: React.ReactNode;
}

export const Button = ({ children, onClick }: ButtonProps) => {
  return <button onClick={() => onClick && onClick()}>{children}</button>;
};
