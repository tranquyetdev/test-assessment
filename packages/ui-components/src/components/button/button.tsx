import * as React from 'react';

interface BaseButtonProps {
  type: 'button' | 'link';
  children: React.ReactNode;
  LeftIcon?: React.ComponentType<any>;
  RightIcon?: React.ComponentType<any>;
  className?: string;
}

interface NormalButtonProps extends BaseButtonProps {
  type: 'button';
  onClick: () => void;
}

interface LinkButtonProps extends BaseButtonProps {
  type: 'link';
  href: string;
  title: string;
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'];
}

export type ButtonProps = NormalButtonProps | LinkButtonProps;

// TODO: apply design
const iconClassName = '';

// TODO: apply design
const buttonClassName: Record<BaseButtonProps['type'], string> = {
  button: '',
  link: '',
};

export const Button = React.forwardRef((props: ButtonProps, ref) => {
  const { children, type, className, LeftIcon, RightIcon } = props;

  if (type === 'link') {
    return (
      <a
        // @ts-expect-error due to unknown type of ref
        ref={ref}
        className={`${buttonClassName[type]} ${className}`}
        href={props.href}
        title={props.title}
        target={props.target || '_blank'}
      >
        {LeftIcon && <LeftIcon className={iconClassName} />}

        {children}

        {RightIcon && <RightIcon className={iconClassName} />}
      </a>
    );
  }

  return (
    <button
      // @ts-expect-error due to unknown type of ref
      ref={ref}
      className={`${buttonClassName[type]} ${className}`}
      onClick={props.onClick}
    >
      {LeftIcon && <LeftIcon className={iconClassName} />}

      {children}

      {RightIcon && <RightIcon className={iconClassName} />}
    </button>
  );
});
