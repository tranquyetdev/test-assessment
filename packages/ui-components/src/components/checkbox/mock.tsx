import { CheckBoxProps, defaultBgcolor } from '.';

export const mockCheckBoxProps: CheckBoxProps = {
  label: 'Developer',
  bgColor: defaultBgcolor,
  id: 'checkbox',
  checked: false,
  disabled: false,
};
