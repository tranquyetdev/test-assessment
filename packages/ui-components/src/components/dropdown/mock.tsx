export const mockDropdownProps = {
  content: (
    <ul className="menu">
      <li className="menu-item">
        <button>Menu 1</button>
      </li>
      <li className="menu-item">
        <button>Menu 2</button>
      </li>
    </ul>
  ),
  button: <div className="border">Show content</div>,
};
