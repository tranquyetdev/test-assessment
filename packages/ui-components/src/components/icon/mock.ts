import { IconName, IconProps } from ".";

export const mockIcons: IconProps & { name?: IconName }[] = [
  { name: "arrow-left" },
  { name: "arrow-right" },
  { name: "check" },
  { name: "chevron-down" },
  { name: "remove" },
  { name: "edit" },
  { name: "notification" },
  { name: "plus" },
  { name: "publish" },
  { name: "qa" },
  { name: "refresh" },
  { name: "save" },
  { name: "search" },
  { name: "caution" },
  { name: "eye-hide" },
  { name: "eye-show" },
]