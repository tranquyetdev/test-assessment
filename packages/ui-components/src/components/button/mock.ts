import { ButtonProps } from './button';

export const mockButtonProps: Record<ButtonProps['type'], ButtonProps> = {
  button: {
    type: 'button',
    children: 'button',
    onClick: () => console.log('hello world'),
  },
  link: {
    type: 'link',
    children: 'link',
    href: 'https://github.com/',
    title: 'Github',
  },
};
