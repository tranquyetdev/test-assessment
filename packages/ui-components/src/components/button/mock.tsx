import { RenderIcon } from '../icons';
import { ButtonProps } from './button';

export const mockButtonProps: Record<ButtonProps['type'], ButtonProps> = {
  button: {
    type: 'button',
    children: 'button',
    onClick: () => alert('hello world'),
    LeftIcon: <RenderIcon name="caution" className="text-error" />,
    RightIcon: (
      <RenderIcon name="caution" className="text-error" />
    ),
  },
  link: {
    type: 'link',
    children: 'link',
    href: 'https://github.com/',
    title: 'Github',
    LeftIcon: <RenderIcon name="caution" className="text-error" />,
    RightIcon: (
      <RenderIcon name="caution" className="text-error" />
    ),
  },
};
