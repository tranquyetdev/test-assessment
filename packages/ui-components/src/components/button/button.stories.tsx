import type { Meta } from '@storybook/react';
import { Button as ButtonComponent } from './button';
import { mockButtonProps } from './mock';

const Story: Meta<typeof ButtonComponent> = {
  title: 'Components / Button',
  component: ButtonComponent,
};
export default Story;

export const Button = () => (
  <ButtonComponent {...mockButtonProps['button']}>button</ButtonComponent>
);

export const Link = () => (
  <ButtonComponent {...mockButtonProps['link']}>link</ButtonComponent>
);
