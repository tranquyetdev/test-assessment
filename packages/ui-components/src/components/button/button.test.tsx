import { render } from '@testing-library/react';

import { Button } from './button';
import { mockButtonProps } from './mock';

describe('Button', () => {
  Object.entries(mockButtonProps).forEach(([type, props]) => {
    it(`should render ${type} successfully`, () => {
      const { baseElement } = render(<Button {...props} />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
