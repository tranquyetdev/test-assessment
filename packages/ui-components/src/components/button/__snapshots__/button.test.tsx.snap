// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button successfully 1`] = `
<body>
  <div>
    <button
      class="px-4 py-2 bg-primary-base font-medium font-primary text-13 leading-6 text-center  inline-block"
    >
      button
    </button>
  </div>
</body>
`;

exports[`Button should render link successfully 1`] = `
<body>
  <div>
    <a
      class="text-primary-base font-normal font-primary text-15 leading-6 "
      href="https://github.com/"
      target="_blank"
      title="Github"
    >
      link
    </a>
  </div>
</body>
`;
